Harvard dce f/mh 12607 fast composer operations

lkiesow opened the pull request:
This is an updated version of pull request #14 which @phoebem updated [here](https://github.com/lkiesow/opencast/pull/1). I'm fine with this to be merged like this as the reviewer of #14. Note that a few minor fixes of mine are included in this version of the pull request.

This only contains the uncontroversial `demux` and lossless `concat` parts of #14.

Commit:
MH-12607, Demux operation

Demux to deal with multipart files

lkiesow committed:
MH-12607, Fixes for demux operation

This patch mainly cleans up the demux code, updates and fixes the
documentation and collapses some redundant blocks of code.

lkiesow committed:
MH-12607, Remove redundant demux encoder parts

This patch uses the updated process method of the composer's encoder
engine to replace the now redundant parts of the encoder engine
previously used by the demux operation.

Commit:
MH-12607 fast-composer-operations: Lossless Concat

lkiesow committed:
MH-12607, Update lossless concat

This patch updates the concat documentation and fixes a few minor issues
in the new lossless part of the concat code.

gregorydlogan commented on docs/guides/admin/docs/workflowoperationhandlers/concat-woh.md:
```diff
@@ -24,19 +36,57 @@ FFmpeg Filter Example
 Usage
 -----
 
 This operation is quite similar to the compose operation. The only difference is that the input properties are not only
```
Are not only limited?  Or are limited?  The latter is how the rest of the sentence reads.

lkiesow replied:
Old docs :-) 
But I also think this is fine as it is.

gregorydlogan commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/EncoderEngine.java:
```diff
@@ -58,7 +56,7 @@
  * Abstract base class for encoder engines.
  */
-public class EncoderEngine implements AutoCloseable {
+public class encoderEngine implements AutoCloseable {
```
Shouldn't this be capitalised?  I'm surprised checkstyles isn't freaking out right now...

gregorydlogan replied:
That's because it current is freaking out...  Does not build due to checkstyles failing this.

gregorydlogan commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/endpoint/ComposerRestService.java:
```diff
@@ -494,14 +494,16 @@ public Response composite(@FormParam("compositeSize") String compositeSizeJson,
-    @RestParameter(description = "The resolution dimension of the concat video as JSON", isRequired = false, name = "outputDimension", type = Type.STRING),
-    @RestParameter(description = "The  frame rate of the concat video (should be positive, e.g. 25.0). Negative values and zero will deactivate frame rate operation.",
-            isRequired = false, name = "outputFrameRate", type = Type.STRING)}, reponses = {
+          @RestParameter(description = "The resolution dimension of the concat video as JSON", isRequired = false, name = "outputDimension", type = Type.STRING),
```
A callout, or a (very) brief example would be nice here.  Are we talking "dimension: 800x600"?  Or { height: 800, width: 600}?  A link to the docs would be good here.

lkiesow replied:
True. I'm going to ignore this for now though since it's unmodified documentation ;-P

gregorydlogan commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/endpoint/ComposerRestService.java:
```diff
@@ -494,14 +494,16 @@ public Response composite(@FormParam("compositeSize") String compositeSizeJson,
-    @RestParameter(description = "The  frame rate of the concat video (should be positive, e.g. 25.0). Negative values and zero will deactivate frame rate operation.",
-            isRequired = false, name = "outputFrameRate", type = Type.STRING)}, reponses = {
+          @RestParameter(description = "The resolution dimension of the concat video as JSON", isRequired = false, name = "outputDimension", type = Type.STRING),
+          @RestParameter(description = "The  frame rate of the concat video (should be positive, e.g. 25.0). Negative values and zero will deactivate frame rate operation.",
```
What does deactivating frame rate operation mean?

gregorydlogan commented on modules/composer-ffmpeg/src/main/java/org/opencastproject/composer/impl/endpoint/ComposerRestService.java:
```diff
@@ -629,6 +631,46 @@ public Response convertImage(@FormParam("sourceImage") String sourceImageXml, @F
   }
 
+  /**
+   * #DCE OPC-29 Demuxes a track into multiple outputs
```
Ideally we don't include private ticket numbers in upstream source contributions.  Likewise, this method is missing its description.

gregorydlogan commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ConcatWorkflowOperationHandler.java:
```diff
@@ -192,39 +209,42 @@ private WorkflowOperationResult concat(MediaPackage src, WorkflowOperationInstan
+      if (outputResolution.startsWith(OUTPUT_PART_PREFIX)) {
+        if (!trackSelectors.keySet().contains(
+            Integer.parseInt(outputResolution.substring(OUTPUT_PART_PREFIX.length()))))
+          throw new WorkflowOperationException("Output resolution part not set!");
```
Incorrect error message.  In this case it's not *found*.  It may or may not be set.  Ideally we check if the prefix is set earlier.

gregorydlogan replied:
I'm also somewhat confused what we're doing here.

If we have (potentially) mismatched inputs
if our output resoltion starts with the output partition prefix?
if the track selector keyset does not contain a parsable integer within the output resolution with the partition prefix removed?
throw an exception

What's the goal?  And are the variables messed up or am I misreading things?

gregorydlogan commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ConcatWorkflowOperationHandler.java:
```diff
@@ -192,39 +209,42 @@ private WorkflowOperationResult concat(MediaPackage src, WorkflowOperationInstan
+        if (!NumberUtils.isCreatable(outputFrameRate.substring(OUTPUT_PART_PREFIX.length()))
                 || !trackSelectors.keySet().contains(Integer.parseInt(
                         outputFrameRate.substring(OUTPUT_PART_PREFIX.length())))) {
           throw new WorkflowOperationException("Output frame rate part not set or invalid!");
```
It would be helpful to know which one is bad.  Also, what's the format here?  I'm guessing something like part-n-25, which isn't going to format correctly with the given code.  This format does not appear to be in the docs as far as I can see.

gregorydlogan commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/ConcatWorkflowOperationHandler.java:
```diff
@@ -246,15 +266,24 @@ private WorkflowOperationResult concat(MediaPackage src, WorkflowOperationInstan
-                "No track has been found with flavor '{}' and/or tag '{}' for concat operation, skipping concatenation!",
-                currentFlavor, currentTag);
+        logger.warn("No track has been found with flavor '{}' and/or tag '{}' for concat operation, "
+                        + "skipping concatenation!", currentFlavor, currentTag);
```
needs a space between the quote and skipping.

gregorydlogan commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/DemuxWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,261 @@
+  @Override
+  public WorkflowOperationResult start(final WorkflowInstance workflowInstance, JobContext context)
+          throws WorkflowOperationException {
+    logger.debug("Running DCE DEmux workflow operation on workflow {}", workflowInstance.getId());
```
Debug message should be rephrased.

gregorydlogan commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/DemuxWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,261 @@
+import java.util.Map;
+
+/**
+ * The workflow definition for handling "Demux from epiphan CA" operations This allows to demux and tagging to be done
```
This should be rephrased for inclusion upstream

lkiesow committed:
Fixed typo

gregorydlogan commented on modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/DemuxWorkflowOperationHandler.java:
```diff
@@ -0,0 +1,261 @@
+      }
+
+      List<Track> composedTracks = (List<Track>) MediaPackageElementParser.getArrayFromXml(job.getPayload());
+      if (composedTracks.size() != targetFlavors.size() && targetFlavors.size() != 1) {
```
Shouldn't we be checking if targetFlavors.size() is not 1 prior to dispatching the jobs?

lkiesow replied:
No, that's ok. if it is 1, the same target flavor will be used for all outputs

gregorydlogan commented:
I'm out of time for today, will continue review tomorrow.

lkiesow committed:
MH-12607, Fixed minor issues

This patch fixes a few minor issues identified during the review.

lkiesow commented:
I'm declining this in favor of #171 and #172 which separate both features contained in this pull request to make the review easier. @gregorydlogan, I've also addressed your comments from yesterday.
